src/snarks/groth16/mpc/phase2.hpp

```diff
@@ -0,0 +1,309 @@
+    const srs_mpc_phase2_challenge<ppT> &challenge,
+    const libff::Fr<ppT> &delta_j);
+
+/// Verify a response against a given challenge.  Checks that the response
+/// matches the expected hash in the challenge, and leverages
+/// `srs_mpc_phase2_verify_update` to validate the claimed contribution.
+template<typename ppT>
+bool srs_mpc_phase2_verify_response(
+    const srs_mpc_phase2_challenge<ppT> &challenge,
+    const srs_mpc_phase2_response<ppT> &response);
+
+/// Given a `response` (which should already have been validated with
+/// `srs_mpc_phase2_verify_response`), create a new challenge object.  This
+/// essentially copies the accumulator, and updates the transcript digest for
+/// the next contributor.
+template<typename ppT>
+srs_mpc_phase2_challenge<ppT> srs_mpc_phase2_compute_challenge(
+    srs_mpc_phase2_response<ppT> &&response);
+
+/// The transcript of the MPC is formed of a sequence of contributions (public
+/// key).  Each contribution contains the digest of the previous one, and the
+/// value of delta after it has been applied.  Therefore the final accumulator
+/// can be fully verified using just the transcript as follows:
+///
+/// - establish the initial value of accumulator and the transcript digest
+///
+/// - confirm the transcript of contributions based on this, yielding an
+///   encoding of the final delta
+///
+/// - check that the final accumulator is consistent with the initial
+///   accumulator, based on the final delta
+///
+/// This function validates the transcript as a stream of publickey objects,
+/// outputing the encoding of the final delta in G1.
+template<typename ppT, bool enable_contribution_check = true>
+bool srs_mpc_phase2_verify_transcript(
+    const srs_mpc_hash_t initial_transcript_digest,
+    const libff::G1<ppT> &initial_delta,
+    const srs_mpc_hash_t check_for_contribution,
+    std::istream &transcript_stream,
+    libff::G1<ppT> &out_final_delta,
+    srs_mpc_hash_t out_final_transcript_digest,
+    bool &out_contribution_found);
+
+/// Similar to other transcript verification above, but does not check for the
+/// presence of a specific contribution digest.
+template<typename ppT>
+bool srs_mpc_phase2_verify_transcript(
+    const srs_mpc_hash_t initial_transcript_digest,
+    const libff::G1<ppT> &initial_delta,
+    std::istream &transcript_stream,
+    libff::G1<ppT> &out_final_delta,
+    srs_mpc_hash_t out_final_transcript_digest);
+
+/// Given the output from the first layer of the MPC, perform the 2nd
+/// layer computation using just local randomness for delta. This is not a
+/// substitute for the full MPC with an auditable log of
+/// contributions, but is useful for testing.
+template<typename ppT>
+srs_mpc_phase2_challenge<ppT> srs_mpc_dummy_phase2(
+    const srs_mpc_layer_L1<ppT> &layer1,
+    const libff::Fr<ppT> &delta,
+    size_t num_inputs);
+
+/// Given the output from all phases of the MPC, create the
+/// prover and verification keys for the given circuit.
```

Comment:
`prover and verification keys` -> "proving and verification keys", or "prover and verifier keys"

Reply:
updated